pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java

```diff
@@ -122,166 +119,52 @@ public void runJob() {
-   */
-  protected void onFinish() {
-  }
-
-  /**
-   * Merges the results from the worker threads into a results block.
-   */
-  protected BaseResultsBlock mergeResults()
-      throws Exception {
-    T mergedBlock = null;
-    int numBlocksMerged = 0;
-    long endTimeMs = _queryContext.getEndTimeMs();
-    while (numBlocksMerged < _numOperators) {
-      // Timeout has reached, shouldn't continue to process. `_blockingQueue.poll` will continue to return blocks even
-      // if negative timeout is provided; therefore an extra check is needed
-      long waitTimeMs = endTimeMs - System.currentTimeMillis();
-      if (waitTimeMs <= 0) {
-        return getTimeoutResultsBlock(numBlocksMerged);
-      }
-      BaseResultsBlock blockToMerge = _blockingQueue.poll(waitTimeMs, TimeUnit.MILLISECONDS);
-      if (blockToMerge == null) {
-        return getTimeoutResultsBlock(numBlocksMerged);
-      }
-      if (blockToMerge.getProcessingExceptions() != null) {
-        // Caught exception while processing segment, skip merging the remaining results blocks and directly return the
-        // exception
-        return blockToMerge;
-      }
-      if (mergedBlock == null) {
-        mergedBlock = convertToMergeableBlock((T) blockToMerge);
-      } else {
-        mergeResultsBlocks(mergedBlock, (T) blockToMerge);
-      }
-      numBlocksMerged++;
-      if (isQuerySatisfied(mergedBlock)) {
-        // Query is satisfied, skip merging the remaining results blocks
-        return mergedBlock;
-      }
-    }
-    return mergedBlock;
-  }
-
-  private ExceptionResultsBlock getTimeoutResultsBlock(int numBlocksMerged) {
+  protected ExceptionResultsBlock getTimeoutResultsBlock(int numBlocksMerged) {
     LOGGER.error("Timed out while polling results block, numBlocksMerged: {} (query: {})", numBlocksMerged,
         _queryContext);
     return new ExceptionResultsBlock(QueryException.EXECUTION_TIMEOUT_ERROR,
         new TimeoutException("Timed out while polling results block"));
   }
 
-  /**
-   * Can be overridden for early termination. The input results block might not be mergeable.
-   */
-  protected boolean isQuerySatisfied(T resultsBlock) {
-    return false;
+  @Override
+  public List<Operator> getChildOperators() {
+    return _operators;
   }
 
   /**
-   * Merges a results block into the main mergeable results block.
-   * <p>NOTE: {@code blockToMerge} should contain the result for a segment without any exception. The errored segment
-   * result is already handled.
-   *
-   * @param mergedBlock The block that accumulates previous results. It should be modified to add the information of the
-   *                    other block.
-   * @param blockToMerge The new block that needs to be merged into the mergedBlock.
+   * Start the combine operator.
    */
-  protected abstract void mergeResultsBlocks(T mergedBlock, T blockToMerge);
+  public abstract void start();
```

Comment:
Suggest moving `start()` and `stop()` into `BaseStreamBlockCombine` because they don't apply to single block